File: R/test_gauss_legendre_quadrature.R
#!/usr/bin/env Rscript

library(pracma)
library(rbenchmark)


args = commandArgs(TRUE)

# Get the number of iterations n from the command line.
if (length(args) == 0) {
  stop("At least one argument must be supplied (numIterations).n", call.=FALSE)
} else if (length(args) == 1) {
  n <- as.integer(args[1])
}

#Integrand f(x) to be numerically integrated.
f <- function(x) exp(x)

cat('---------------------------------------\n')
cat('     Gauss Legendre Quadrature: ', n, '\n')
cat('---------------------------------------\n')

#Use Gauss-Legendre quadrature over the interval [a, b] on f(x).
benchmark("Quadrature" = {
            a <- -3.0
            b <-  3.0
            i <- 1:n
            x <- a + (i-1)*(b-a)/(n-1)

            quadrature.values <- gaussLegendre(n, a, b)
            quadrature.result <- sum(quadrature.values$w*f(quadrature.values$x)) 
          },
          replications = 1,
          columns = c("test", "replications", "elapsed", "relative", "user.self", "sys.self"))

File: Python/test_gauss_legendre_quadrature.py
#!/usr/bin/env python

from __future__ import print_function

import numpy as np
import sys
import benchmark_decorator as dectimer

#--------------------
# Function: integrand
#--------------------
integrand = lambda x: np.exp(x)

#-----------------------------
# Function: compute_quadrature
#-----------------------------
@dectimer.bench_time(3)
def compute_quadrature(n):
    """
      Perform the Gauss-Legendre Quadrature at the prescribed order n
    """
    a = -3.0
    b = 3.0

    # Gauss-Legendre (default interval is [-1, 1])
    x, w = np.polynomial.legendre.leggauss(n)

    # Translate x values from the interval [-1, 1] to [a, b]
    t = 0.5*(x + 1)*(b - a) + a

    gauss = sum(w * integrand(t)) * 0.5*(b - a)

if len(sys.argv) < 1:
    print('Usage:')
    print('     python ' + sys.argv[0] + ' N')
    print('Please specify the order of the quadrature.')
    sys.exit()

order = int(sys.argv[1])

print('Gauss-Legendre Quadrature of order: ', order)

compute_quadrature(order)

print(' ')
